### PR TITLE
Templates via rest

### PR DESCRIPTION
**Pull request by lehinevych**
{{pytest: cfme/tests/infrastructure/test_rest_templates.py -v --long-running}}

### CODE REVIEW

**Comment by dajoRH** [comment]
Lint report for commit ec109a0e6fbc77b318a8fd040e3c5f95182fb474:
:hurtrealbad: There were some flake issues that need to be resolved in order to merge the pull request:
```
./cfme/tests/infrastructure/test_rest_templates.py:9:1: E302 expected 2 blank lines, found 1
./cfme/tests/infrastructure/test_rest_templates.py:53:5: E113 unexpected indentation
./cfme/tests/infrastructure/test_rest_templates.py:53:5: E901 IndentationError: unexpected indent
```
:trollface:
*CFME QE Bot*

**Comment by dajoRH** [comment]
Lint report for commit 8bfcfe02e3353477d26b46bcf74842a9c4af0607:
:hurtrealbad: There were some flake issues that need to be resolved in order to merge the pull request:
```
./cfme/tests/infrastructure/test_rest_templates.py:54:5: E113 unexpected indentation
./cfme/tests/infrastructure/test_rest_templates.py:54:5: E901 IndentationError: unexpected indent
```
:trollface:
*CFME QE Bot*

**Comment by dajoRH** [comment]
Lint report for commit 376955024549eea74fb0c30ea8ddf56fc92751cd:
:godmode: All seems good! :cake: :punch: :cookie: 
*CFME QE Bot*

**Comment by dajoRH** [comment]
Lint report for commit 019eb97d8f519dde7e2e0b92d6504c59cef9c5eb:
:godmode: All seems good! :cake: :punch: :cookie: 
*CFME QE Bot*

**Comment by dajoRH** [comment]
Lint report for commit 804ea0c9a35aff79b335874c362e334ce769fe32:
:godmode: All seems good! :cake: :punch: :cookie: 
*CFME QE Bot*

File: cfme/rest/__init__.py [hunk 1]
```diff
@@ -314,3 +316,28 @@ def automation_requests_data(vm):
...
+    for tpl in template_names:
+        wait_for(lambda: rest_api.collections.templates.find_by(name=tpl), num_sec=180, delay=10)
+        tpls.append(rest_api.collections.templates.get(name.tpl))
+
+    @request.addfinalizer
+    def _finished():
+        ids = [t.id for t in tpls]
+        delete_tpls = [t for t in rest_api.collections.templates if t.id in ids]
+        if len(delete_tpls) != 0:
+            rest_api.collections.tempaltes.action.delete(*delete_tpls)
```
Comment (mfalesni) [inline cfme/rest/__init__.py]: I see a typo here.

**Comment by dajoRH** [comment]
Would you mind rebasing this Pull Request against latest master, please? :trollface:
*CFME QE Bot*

**Comment by dajoRH** [comment]
Lint report for commit 36cada1c9e0fab8218a73f2bd6f1cc1664e37e02:
:godmode: All seems good! :cake: :punch: :cookie: 
*CFME QE Bot*

**Comment by dajoRH** [comment]
Lint report for commit 4d5765670549c6c66ad1e68b9a7c4133862d3c17:
:godmode: All seems good! :cake: :punch: :cookie: 
*CFME QE Bot*

**Comment by lehinevych** [comment]
In the template fixture  I use method  publish_to_template (cfme/infrastructure/virtual_machines.py:357)
however on RHEV 3.4 this option isn't available for vm. The question is how it will be better to handler this case or maybe there is more efficient way to create new template for test?

File: cfme/rest/__init__.py [hunk 2]
```diff
@@ -339,3 +341,28 @@ def _finished():
...
+        tpls.append(rest_api.collections.templates.get(name.tpl))
+
+    @request.addfinalizer
+    def _finished():
+        ids = [t.id for t in tpls]
+        delete_tpls = [t for t in rest_api.collections.templates if t.id in ids]
+        if len(delete_tpls) != 0:
+            rest_api.collections.templates.action.delete(*delete_tpls)
+
+    if num == 1:
```
Comment (RonnyPfannschmidt) [inline cfme/rest/__init__.py]: this destroys finalization due to the `pop`
it also creates a variadic result type

i suggest to always return a list and unpack at the usage site
that way the return type is consistent and the finalization can always run

Comment (lehinevych) [inline cfme/rest/__init__.py]: @RonnyPfannschmidt 
Agree, however I think it will be better to return tpls[-1]?

Comment (RonnyPfannschmidt) [inline cfme/rest/__init__.py]: true, it seems that is a consistent pattern on `num=1`

**Comment by dajoRH** [comment]
Lint report for commit e144ab147e6080d937924cf69b47232e9a742183:
:godmode: All seems good! :cake: :punch: :cookie: 
*CFME QE Bot*

**Comment by dajoRH** [comment]
Lint report for commit 661e1df0aaad66a8da872f790e41651a027d030d:
:godmode: All seems good! :cake: :punch: :cookie: 
*CFME QE Bot*

**Comment by dajoRH** [comment]
Lint report for commit 038cf4aa2bc6162d92d926de4747231b19afb89a:
:godmode: All seems good! :cake: :punch: :cookie: 
*CFME QE Bot*

**Comment by dajoRH** [comment]
Lint report for commit 305ce077f92afaa426c5ed45ad66e53975eeaceb:
:godmode: All seems good! :cake: :punch: :cookie: 
*CFME QE Bot*

**Comment by dajoRH** [comment]
Lint report for commit 782256924c14ed5d4960d5945dce998fb8622e03:
:godmode: All seems good! :cake: :punch: :cookie: 
*CFME QE Bot*

**Comment by dajoRH** [comment]
Lint report for commit 59cadbe04dda8ca836ed173c479a87ec8c876def:
:godmode: All seems good! :cake: :punch: :cookie: 
*CFME QE Bot*

**Comment by dajoRH** [comment]
Lint report for commit ca9a50258df3b5c9336b7542aac6c4e3131eb67f:
:godmode: All seems good! :cake: :punch: :cookie: 
*CFME QE Bot*

**Comment by dajoRH** [comment]
Lint report for commit 7111d3f7c49a03bd813a669327a6c8ed476a057c:
:godmode: All seems good! :cake: :punch: :cookie: 
*CFME QE Bot*

File: cfme/rest/__init__.py [hunk 3]
```diff
@@ -339,3 +340,30 @@ def _finished():
             collection.action.delete(*delete_entities)
 
     return entities
+
+
+def templates(request, rest_api, a_provider, vm, num=1):
```
Comment (psav) [inline cfme/rest/__init__.py]: I think there a few things I'd change here
1) This function's name is misleading, it should probably be called, ```templatize_vm``` or something else, please add a docstring to it also
2) It should be split up between two functions, there should be a fixture in the ```test``` file. We shouldn't require a ```request``` object and definitely shouldn't ```addfinalizer``` inside this module.
3) In splitting it up, the iteration can happen in the fixture, not in this function.

**Comment by dajoRH** [comment]
Lint report for commit 79af2dfddb6c6602164f300b9cfe327ff6b47909:
:godmode: All seems good! :cake: :punch: :cookie: 
*CFME QE Bot*

**Comment by dajoRH** [comment]
Lint report for commit 6f547fe1db8befa9f1dad65e3bd384542108300c:
:godmode: All seems good! :cake: :punch: :cookie: 
*CFME QE Bot*

**Comment by dajoRH** [comment]
Lint report for commit 926f17d1cb9bc0973891e924b22f871e17ccc113:

Everything seems all right :smile: .
*CFME QE Bot*

**Comment by dajoRH** [comment]
Lint report for commit 56a8e9c2a63ceef9b38779c3982415dd628c9ef5:

Everything seems all right :smile: .
*CFME QE Bot*

**Comment by dajoRH** [comment]
Lint report for commit 2ea74546d1a78e2ea2486eabb04ffb1ae87c28b0:

Everything seems all right :smile: .
*CFME QE Bot*

File: cfme/rest/__init__.py [hunk 4]
```diff
@@ -338,3 +339,21 @@ def _finished():
...
+
+def mark_vm_as_template(rest_api, provider, vm_name):
+    """
+        Function marks vm as template via mgmt and returns template Entity
+        Usage:
+            mark_vm_as_template(rest_api, provider, vm_name)
+    """
+    t_vm = rest_api.collections.vms.get(name=vm_name)
+    t_vm.action.stop()
+    wait_for(lambda: t_vm.power_state == "off", num_sec=600, delay=15, fail_func=t_vm.reload)
```
Comment (mfalesni) [inline cfme/rest/__init__.py]: I think you can use wait_vm_stopped from the provider mgmt.

**Comment by dajoRH** [comment]
Lint report for commit 7f06f2a4faadb89eba194cb9ba2b9591494b2aea:

Everything seems all right :smile: .
*CFME QE Bot*

**Comment by lehinevych** [comment]
Fails on upstream because of APIException: JSONDecodeError: 500 Internal Server Error

File: cfme/tests/infrastructure/test_rest_templates.py [hunk 5]
```diff
@@ -0,0 +1,71 @@
...
+
+    @request.addfinalizer
+    def _finished():
+        if template.id in rest_api.collections.templates:
+            rest_api.collections.templates.action.delete(*template)
+
+    return template
+
+
+@pytest.mark.uncollectif(lambda: version.current_version() < '5.5')
```
Comment (jkrocil) [inline cfme/tests/infrastructure/test_rest_templates.py]: We already have `pytest.mark.ignore_stream("5.4")` in this module so we dont need this line.

Comment (jkrocil) [inline cfme/tests/infrastructure/test_rest_templates.py]: Otherwise LGTM.

Comment (lehinevych) [inline cfme/tests/infrastructure/test_rest_templates.py]: Removed

**Comment by dajoRH** [comment]
Lint report for commit 2b4d8f84b73905bb70730c2e8ef482e171a79458:

`cfme/rest/__init__.py`:
- File lint OK :cake: :punch: :cookie:

`cfme/tests/infrastructure/test_rest_templates.py`:
- :red_circle: Line 7:1: **F401** *'version' imported but unused*

Please, rectify these issues :smirk: .
*CFME QE Bot*

**Comment by dajoRH** [comment]
Lint report for commit 519966ff04486858991b93b0c0ccca34ac303f37:

Everything seems all right :smile: .
*CFME QE Bot*

**Comment by dajoRH** [comment]
Lint report for commit ad5834e6e6b77631b56d2aa313de188158b5a8fb:

Everything seems all right :smile: .
*CFME QE Bot*

**Comment by jkrocil** [comment]
Merging.

**Pull request merged**